pyquil/paulis.py

```diff
@@ -203,13 +203,13 @@ def _multiply_factor(self, factor, index):
 
         return new_term
 
-    def __mul__(self, term):
+    def __mul__(self, term: Union['PauliTerm', 'PauliSum', Number]) -> Union['PauliTerm', 'PauliSum']:
```

Comment:
Ultimately, we'll probably want to convert this to use [function overloading](https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading) to help mypy work out the relationship between `term`'s type and the return type, since `Union` return types are [sometimes cumbersome](https://github.com/python/mypy/issues/1693).

But I think it's fine to leave it as is for now, and address that as this module gets closer to being fully typed